# Connect to database

In [1]:
import os
import subprocess
import requests
#start the database
process = subprocess.Popen(["postgres", "-D", "/home/mcn26/palmer_scratch/db", "-p", "5433"])
# Load the SQL extension
%load_ext sql
#dealing with a weird bug
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

2025-04-01 10:08:19.062 EDT [623798] FATAL:  could not open directory "pg_notify": No such file or directory
2025-04-01 10:08:19.069 EDT [623798] LOG:  database system is shut down


In [2]:
%%sql
postgresql://mr_root:password@localhost:5433/scratch

Traceback (most recent call last):
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_varef/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 146, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_varef/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3302, in raw_connection
    return self.pool.connect()
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_varef/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 449, in connect
    return _ConnectionFairy._checkout(self)
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_varef/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 1263, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_varef/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 712, in checkout
    rec = pool._do_get()
  File "/gpfs/gibbs/project/reilly/mcn26/conda_envs/mcn_vare

Our emvar definition is abs(skew)>=0.5 and max(alt activity, ref activity)>=1

In [3]:
%%sql
SELECT *
FROM stacked_table
LIMIT 5;

 * postgresql://mr_root:***@localhost:5433/scratch
5 rows affected.


[('AFR', 'rs1000778', 33462484, 'Genetics of 35 blood and urine biomarkers in the UK Biobank.', 'Apolipoprotein A1 levels', '4E-14', 'rs174456', '0.99', 'chr11', 61888710, 'C', 'A', 0.2905388, 0.47739273, 0.25654978, 0.4235628, 0.68201435, 0.30011693, 0.13302402, 0.20462163, 0.04356711, None, None, None, None, None),
 ('AFR', 'rs1000778', 33462484, 'Genetics of 35 blood and urine biomarkers in the UK Biobank.', 'Apolipoprotein A1 levels', '4E-14', 'rs174457', '0.99', 'chr11', 61889503, 'T', 'C', 1.2269007, 0.87899405, 1.3708335, 1.1865149, 0.88644683, 1.3176237, -0.040385775, 0.007452801, -0.053209677, None, None, None, None, None),
 ('AFR', 'rs1000778', 33462484, 'Genetics of 35 blood and urine biomarkers in the UK Biobank.', 'Apolipoprotein A1 levels', '4E-14', 'rs174458', '0.8', 'chr11', 61889523, 'C', 'T', 1.1304642, 0.95638263, 1.4404697, 1.1269263, 0.9378151, 1.3677735, -0.0035378272, -0.018567348, -0.07269636, None, None, None, None, None),
 ('AFR', 'rs1000778', 33462484, 'Genetics of 35 blood and urine biomarkers in the UK Biobank.', 'Apolipoprotein A1 levels', '4E-14', 'rs174459', '0.94', 'chr11', 61889578, 'C', 'T', 1.1565859, 1.3619026, 1.8750244, 0.93656266, 1.1710726, 1.5803186, -0.22002321, -0.19082987, -0.29470596, None, None, None, None, None),
 ('AFR', 'rs1000778', 33462484, 'Genetics of 35 blood and urine biomarkers in the UK Biobank.', 'Apolipoprotein A1 levels', '4E-14', 'rs174460', '0.87', 'chr11', 61889638, 'C', 'T', 1.0319768, 1.3778319, 1.8516933, 1.0674824, 1.379455, 1.8930544, 0.03550557, 0.0016230709, 0.04136129, None, None, None, None, None)]

In [9]:
%%sql
CREATE TABLE stacked_table_deduplicated AS
SELECT DISTINCT source_table, tag_snp, pubmedid, study, disease_trait, p_value, ld_buddy, r2, chromosome_id, pos, ref, alt, k562__ref, hepg2__ref, sknsh__ref, k562__alt, hepg2__alt, sknsh__alt, k562__skew, hepg2__skew, sknsh__skew
FROM stacked_table;

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 10:00:32.035 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:00:41.446 EDT [2811185] LOG:  checkpoint complete: wrote 147 buffers (0.9%); 0 WAL file(s) added, 1 removed, 32 recycled; write=8.655 s, sync=0.520 s, total=9.412 s; sync files=29, longest=0.333 s, average=0.018 s; distance=528689 kB, estimate=528689 kB; lsn=10E/245DE838, redo lsn=10E/54085B0
2025-03-18 10:00:41.668 EDT [2811185] LOG:  checkpoints are occurring too frequently (9 seconds apart)
2025-03-18 10:00:41.668 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:00:41.676 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:00:50.162 EDT [2811185] LOG:  checkpoint complete: wrote 11 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=8.062 s, sync=0.211 s, total=8.494 s; sync files=5, longest=0.191 s, average=0.043 s; distance=537478 kB, estimate=537478 kB; lsn=10E/4491B108, redo lsn=10E/260EA040
2025-03-18 10:00:50.794 EDT [281118

16412995 rows affected.


[]

In [10]:
%%sql
SELECT 'undedup' AS name, COUNT(*) AS rowcount 
FROM stacked_table
UNION ALL
SELECT 'dedup' AS name, COUNT(*) AS rowcount 
FROM stacked_table_deduplicated;

 * postgresql://mr_root:***@localhost:5433/scratch
2 rows affected.


[('dedup', 16412995), ('undedup', 32825990)]

That is a LOT of duplicates. Where are they coming from? I *guess* that they are from the LD expansion step, but I don't know that. Let's take a closer look.

Examining the original tables loaded in the load_gwas step, I see the difference between deduplicated & undeduplicated is:
- AFR: 11103
- ASN: 22306
- EUR: 93255
- SUM: 126664

Which is a lot less than the difference observed above
- 16412995

This does make sense considering that every GWAS row probably has multiple predictions... Let's check if our initial loaded tables match...

In [3]:
%%sql
SELECT *
FROM gwas_AFR
LIMIT 5;

 * postgresql://mr_root:***@localhost:5433/scratch
5 rows affected.


[('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs1000005', 'tag'),
 ('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs7275426', '0.83'),
 ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs10000225', 'tag'),
 ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs10009710', '0.71'),
 ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs12640763', '0.74')]

2025-03-18 10:51:48.947 EDT [3107781] LOG:  checkpoint starting: time
2025-03-18 10:53:14.615 EDT [3107781] LOG:  checkpoint complete: wrote 343 buffers (2.1%); 0 WAL file(s) added, 6 removed, 0 recycled; write=85.112 s, sync=0.512 s, total=85.668 s; sync files=21, longest=0.334 s, average=0.025 s; distance=94972 kB, estimate=94972 kB; lsn=11C/B6BCB8C8, redo lsn=11C/B48FB088


In [6]:
%%sql
SELECT 'undedup' AS name, COUNT(*) AS rowcount 
FROM gwas_AFR
UNION ALL
SELECT 'dedup' AS name, COUNT(*) AS rowcount 
FROM (
    SELECT DISTINCT rsid, pubmedid, study, disease_trait, p_value, broad_ancestral_category, ld_buddy, r2
    FROM gwas_AFR
) AS deduped;

 * postgresql://mr_root:***@localhost:5433/scratch
2 rows affected.


[('undedup', 3026426), ('dedup', 1502110)]

2025-03-18 11:16:48.694 EDT [3107781] LOG:  checkpoint starting: time
2025-03-18 11:17:05.193 EDT [3113268] LOG:  automatic vacuum of table "scratch.public.stacked_table_deduplicated": index scans: 0
	pages: 0 removed, 3636832 remain, 3465035 scanned (95.28% of total)
	tuples: 77723 removed, 17226754 remain, 0 are dead but not yet removable
	removable cutoff: 936, which was 0 XIDs old when operation ended
	new relfrozenxid: 935, which is 13 XIDs ahead of previous value
	frozen: 36963 pages from table (1.02% of total) had 36986 tuples frozen
	index scan not needed: 0 pages from table (0.00% of total) had 0 dead item identifiers removed
	avg read rate: 15.962 MB/s, avg write rate: 15.580 MB/s
	buffer usage: 3465291 hits, 3465948 misses, 3383017 dirtied
	WAL usage: 6156375 records, 2774337 full page images, 427830277 bytes
	system usage: CPU: user: 6.43 s, system: 20.55 s, elapsed: 1696.37 s
2025-03-18 11:17:59.708 EDT [3107781] LOG:  checkpoint complete: wrote 191 buffers (1.2%); 0 WAL f

So for AFR sql tab: 3026426-1502110=1524316

uhhhhhhhhhh
Original file has 
- 1513214 total
- 1502111 dedup

So the dedup at least matches ! (off-by-one because original file counts header as a line)

# Add boolean emvar columns

In [13]:
%%sql
--ALTER TABLE stacked_table_deduplicated
--DROP COLUMN k562_emvar; 
ALTER TABLE stacked_table_deduplicated ADD COLUMN k562_emvar BOOLEAN;
UPDATE stacked_table_deduplicated
SET k562_emvar = (
    CASE
        WHEN GREATEST(k562__ref, k562__alt) >= 1 AND ABS(k562__skew) > 0.5 THEN TRUE
        ELSE FALSE
    END
);

 * postgresql://mr_root:***@localhost:5433/scratch
Done.


2025-03-18 10:07:11.387 EDT [2811185] LOG:  checkpoint complete: wrote 181 buffers (1.1%); 0 WAL file(s) added, 0 removed, 24 recycled; write=20.277 s, sync=1.146 s, total=21.584 s; sync files=26, longest=0.795 s, average=0.045 s; distance=399839 kB, estimate=528257 kB; lsn=10F/49910ED8, redo lsn=10F/25A79C70
2025-03-18 10:07:11.387 EDT [2811185] LOG:  checkpoints are occurring too frequently (22 seconds apart)
2025-03-18 10:07:11.387 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:07:11.387 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:07:14.817 EDT [2811185] LOG:  checkpoint complete: wrote 186 buffers (1.1%); 0 WAL file(s) added, 7 removed, 31 recycled; write=2.748 s, sync=0.470 s, total=3.431 s; sync files=9, longest=0.236 s, average=0.053 s; distance=618138 kB, estimate=618138 kB; lsn=10F/6B401860, redo lsn=10F/4B6207E0
2025-03-18 10:07:14.828 EDT [2811185] LOG:  checkpoints are occurring too frequently (3 seconds apa

16412995 rows affected.


[]

In [14]:
%%sql
ALTER TABLE stacked_table_deduplicated ADD COLUMN hepg2_emvar BOOLEAN;
UPDATE stacked_table_deduplicated
SET hepg2_emvar = (
    CASE
        WHEN GREATEST(hepg2__ref, hepg2__alt) >= 1 AND ABS(hepg2__skew) > 0.5 THEN TRUE
        ELSE FALSE
    END
);

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 10:08:15.243 EDT [2974776] ERROR:  canceling autovacuum task
2025-03-18 10:08:15.243 EDT [2974776] CONTEXT:  while scanning block 112053 of relation "public.stacked_table_deduplicated"
	automatic vacuum of table "scratch.public.stacked_table_deduplicated"


Done.


2025-03-18 10:08:24.948 EDT [2811185] LOG:  checkpoint complete: wrote 219 buffers (1.3%); 0 WAL file(s) added, 25 removed, 11 recycled; write=6.929 s, sync=4.330 s, total=11.476 s; sync files=14, longest=0.863 s, average=0.310 s; distance=591164 kB, estimate=591164 kB; lsn=111/F84026F8, redo lsn=111/C454FBF0
2025-03-18 10:08:24.982 EDT [2811185] LOG:  checkpoints are occurring too frequently (11 seconds apart)
2025-03-18 10:08:24.982 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:08:24.992 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:08:29.226 EDT [2811185] LOG:  checkpoint complete: wrote 203 buffers (1.2%); 0 WAL file(s) added, 23 removed, 31 recycled; write=3.279 s, sync=0.495 s, total=4.245 s; sync files=8, longest=0.286 s, average=0.062 s; distance=895539 kB, estimate=895539 kB; lsn=112/1A403270, redo lsn=111/FAFDC8B0
2025-03-18 10:08:29.243 EDT [2811185] LOG:  checkpoints are occurring too frequently (5 seconds ap

16412995 rows affected.


[]

In [15]:
%%sql
ALTER TABLE stacked_table_deduplicated ADD COLUMN sknsh_emvar BOOLEAN;
UPDATE stacked_table_deduplicated
SET sknsh_emvar = (
    CASE
        WHEN GREATEST(sknsh__ref, sknsh__alt) >= 1 AND ABS(sknsh__skew) > 0.5 THEN TRUE
        ELSE FALSE
    END
);

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 10:09:28.804 EDT [2983990] ERROR:  canceling autovacuum task
2025-03-18 10:09:28.804 EDT [2983990] CONTEXT:  while scanning block 14403 of relation "public.stacked_table_deduplicated"
	automatic vacuum of table "scratch.public.stacked_table_deduplicated"


Done.


2025-03-18 10:09:40.984 EDT [2811185] LOG:  checkpoint complete: wrote 288 buffers (1.8%); 0 WAL file(s) added, 15 removed, 22 recycled; write=12.699 s, sync=1.862 s, total=14.841 s; sync files=20, longest=0.934 s, average=0.094 s; distance=597335 kB, estimate=622897 kB; lsn=114/9919EBA8, redo lsn=114/70179680
2025-03-18 10:09:40.999 EDT [2811185] LOG:  checkpoints are occurring too frequently (14 seconds apart)
2025-03-18 10:09:40.999 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:09:41.000 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:09:44.574 EDT [2811185] LOG:  checkpoint complete: wrote 1095 buffers (6.7%); 0 WAL file(s) added, 12 removed, 32 recycled; write=2.882 s, sync=0.456 s, total=3.576 s; sync files=8, longest=0.146 s, average=0.057 s; distance=723500 kB, estimate=723500 kB; lsn=114/BB424A28, redo lsn=114/9C4047A8
2025-03-18 10:09:44.611 EDT [2811185] LOG:  checkpoints are occurring too frequently (4 seconds 

16412995 rows affected.


[]

In [16]:
%%sql
--checking a few entries to make sure they make sense. 
SELECT *
FROM stacked_table_deduplicated
WHERE k562_emvar=True
LIMIT 2;

 * postgresql://mr_root:***@localhost:5433/scratch
2 rows affected.


[('AFR', 'rs1077215', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '2E-10', 'rs113251786', '0.82', 'chr3', 50676729, 'G', 'T', 2.666311, 1.7083129, 1.4569813, 3.1763575, 1.9124666, 1.6652815, 0.5100464, 0.20415378, 0.2083003, True, False, False),
 ('AFR', 'rs10795395', 36635386, 'Genomic atlas of the plasma metabolome prioritizes metabolites implicated in human diseases.', 'N-acetyltaurine levels', '6E-50', 'rs7922050', '0.81', 'chr10', 16437008, 'C', 'G', 4.2308292, 3.7484484, 4.084931, 3.6644473, 3.2705476, 3.6303384, -0.5663817, -0.47790056, -0.45459256, True, False, False)]

# Add aggregate emvar columns

In [17]:
%%sql
ALTER TABLE stacked_table_deduplicated ADD COLUMN any_emvar BOOLEAN;
UPDATE stacked_table_deduplicated
SET any_emvar = (
    k562_emvar OR hepg2_emvar OR sknsh_emvar
);

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 10:10:49.650 EDT [2987906] ERROR:  canceling autovacuum task
2025-03-18 10:10:49.650 EDT [2987906] CONTEXT:  while scanning block 48062 of relation "public.stacked_table_deduplicated"
	automatic vacuum of table "scratch.public.stacked_table_deduplicated"


Done.


2025-03-18 10:11:00.732 EDT [2811185] LOG:  checkpoint complete: wrote 301 buffers (1.8%); 0 WAL file(s) added, 8 removed, 27 recycled; write=10.911 s, sync=1.292 s, total=12.396 s; sync files=24, longest=1.137 s, average=0.054 s; distance=578836 kB, estimate=594505 kB; lsn=117/59911AC8, redo lsn=117/35BDB118
2025-03-18 10:11:00.742 EDT [2811185] LOG:  checkpoints are occurring too frequently (12 seconds apart)
2025-03-18 10:11:00.742 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:11:00.742 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:11:04.166 EDT [2811185] LOG:  checkpoint complete: wrote 180 buffers (1.1%); 0 WAL file(s) added, 6 removed, 32 recycled; write=3.071 s, sync=0.094 s, total=3.425 s; sync files=9, longest=0.052 s, average=0.011 s; distance=623452 kB, estimate=623452 kB; lsn=117/7A404CF8, redo lsn=117/5BCB24C0
2025-03-18 10:11:04.168 EDT [2811185] LOG:  checkpoints are occurring too frequently (4 seconds apa

16412995 rows affected.


[]

In [18]:
%%sql
ALTER TABLE stacked_table_deduplicated ADD COLUMN all_emvar BOOLEAN;
UPDATE stacked_table_deduplicated
SET all_emvar = (
    k562_emvar AND hepg2_emvar AND sknsh_emvar
);

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 10:12:07.323 EDT [2991883] ERROR:  canceling autovacuum task
2025-03-18 10:12:07.323 EDT [2991883] CONTEXT:  while scanning block 96408 of relation "public.stacked_table_deduplicated"
	automatic vacuum of table "scratch.public.stacked_table_deduplicated"


Done.


2025-03-18 10:12:19.010 EDT [2811185] LOG:  checkpoint complete: wrote 322 buffers (2.0%); 0 WAL file(s) added, 2 removed, 32 recycled; write=13.626 s, sync=0.474 s, total=14.410 s; sync files=30, longest=0.309 s, average=0.016 s; distance=552530 kB, estimate=562940 kB; lsn=119/FCA54878, redo lsn=119/DD1B4188
2025-03-18 10:12:19.069 EDT [2811185] LOG:  checkpoints are occurring too frequently (15 seconds apart)
2025-03-18 10:12:19.069 EDT [2811185] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 10:12:19.095 EDT [2811185] LOG:  checkpoint starting: wal
2025-03-18 10:12:23.108 EDT [2811185] LOG:  checkpoint complete: wrote 214 buffers (1.3%); 0 WAL file(s) added, 6 removed, 29 recycled; write=2.971 s, sync=0.797 s, total=4.039 s; sync files=9, longest=0.292 s, average=0.089 s; distance=584449 kB, estimate=584449 kB; lsn=11A/22682960, redo lsn=11A/C748A0
2025-03-18 10:12:23.121 EDT [2811185] LOG:  checkpoints are occurring too frequently (4 seconds apart

16412995 rows affected.


[]

In [19]:
%%sql
--checking a few entries to make sure they make sense. 
--SELECT *
--FROM stacked_table
--WHERE any_emvar=True
--LIMIT 20;

 * postgresql://mr_root:***@localhost:5433/scratch
(psycopg2.ProgrammingError) can't execute an empty query
[SQL: --checking a few entries to make sure they make sense. 
--SELECT *
--FROM stacked_table
--WHERE any_emvar=True
--LIMIT 20;]
(Background on this error at: https://sqlalche.me/e/20/f405)


# Calc unique SNPS which are emvars

In [20]:
%%sql
SELECT *
FROM
    stacked_table_deduplicated
LIMIT 1;

 * postgresql://mr_root:***@localhost:5433/scratch
1 rows affected.


[('AFR', 'rs10797938', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '3E-55', 'rs10797937', '0.97', 'chr1', 184047997, 'A', 'G', 0.3319283, 0.2794782, 0.10045869, 0.39208332, 0.33546147, 0.13487323, 0.060154982, 0.05598326, 0.034414556, False, False, False, False, False)]

In [3]:
%%sql
--Flatten to unique ld_buddy values. 
--DROP TABLE buddy_has_any_emvar;

CREATE TABLE buddy_has_any_emvar AS
SELECT
    ld_buddy,
    
    SUM(any_emvar::INT)>0 AS true_any_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    ld_buddy;

 * postgresql://mr_root:***@localhost:5433/scratch
2041121 rows affected.


[]

In [4]:
%%sql
SELECT 
    COUNT(*) AS total_uniq_ld_buddies,
    SUM(true_any_emvar_count::INT) AS total_emvars
FROM
    buddy_has_any_emvar;

 * postgresql://mr_root:***@localhost:5433/scratch
1 rows affected.


[(2041121, 29130)]

# Calc total associations (source_table X study X disease_trait X tag_snp) with emvars

In [6]:
%%sql
--DROP TABLE association_has_any_emvar;
CREATE TABLE association_has_any_emvar AS
SELECT
    source_table,
    study,
    disease_trait,
    tag_snp,
    SUM(any_emvar::INT)>0 AS true_any_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    source_table,
    study,
    disease_trait,
    tag_snp;

 * postgresql://mr_root:***@localhost:5433/scratch
Done.
759651 rows affected.


[]

In [7]:
%%sql
SELECT 
    COUNT(*) AS total_associations,
    SUM(true_any_emvar_count::INT) AS total_emvars
FROM
    association_has_any_emvar;

 * postgresql://mr_root:***@localhost:5433/scratch
1 rows affected.


[(759651, 163513)]

# Calc total number of associations with some or all emvar

In [8]:
%%sql
SELECT
    source_table,
    COUNT(*) AS total_rows,
    SUM(any_emvar::INT) AS true_any_emvar_count,
    SUM(all_emvar::INT) AS true_all_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    source_table;

 * postgresql://mr_root:***@localhost:5433/scratch
3 rows affected.


[('AFR', 1178968, 18216, 4267),
 ('ASN', 4423038, 66759, 14403),
 ('EUR', 10810989, 169044, 37448)]

# Calculate number of ld buddies, tagsnps, ld buddies X tag snps which have any emvars.

This table, "tag_buddy_combo", reports, for each combination of source_table, tag_snp, ld_buddy, whether that combo has any emvars.

In [9]:
%%sql
--DROP TABLE tag_buddy_combo;
CREATE TABLE tag_buddy_combo AS
SELECT
    source_table,
    ld_buddy,
    tag_snp,
    SUM(any_emvar::INT)>0 AS true_any_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    source_table,
    tag_snp,
    ld_buddy;

 * postgresql://mr_root:***@localhost:5433/scratch
Done.


2025-03-18 13:53:38.342 EDT [847201] LOG:  checkpoint starting: wal
2025-03-18 13:53:50.390 EDT [847201] LOG:  checkpoint complete: wrote 73 buffers (0.4%); 0 WAL file(s) added, 0 removed, 33 recycled; write=11.687 s, sync=0.174 s, total=12.059 s; sync files=46, longest=0.161 s, average=0.004 s; distance=532861 kB, estimate=532861 kB; lsn=11D/52746480, redo lsn=11D/34314A28


9048842 rows affected.


2025-03-18 13:53:51.246 EDT [847201] LOG:  checkpoints are occurring too frequently (13 seconds apart)
2025-03-18 13:53:51.246 EDT [847201] HINT:  Consider increasing the configuration parameter "max_wal_size".
2025-03-18 13:53:51.252 EDT [847201] LOG:  checkpoint starting: wal


[]

Now let's summarize how many combos have any emvars.

In [10]:
%%sql
SELECT
    source_table,
    COUNT(*) AS total_ld_buddy_tag_snp_combos,
    SUM(true_any_emvar_count::INT) AS any_emvar_ld_buddy_tag_snp_combos
FROM 
    tag_buddy_combo
GROUP BY 
    source_table;

 * postgresql://mr_root:***@localhost:5433/scratch
3 rows affected.


[('AFR', 776254, 11112), ('ASN', 2511648, 36296), ('EUR', 5760940, 82563)]

This table, "tags", reports, for each combination of source_table, tag_snp, whether that combo has any emvars.

In [12]:
%%sql
--DROP TABLE tags;
CREATE TABLE tags AS
SELECT
    source_table,
    tag_snp,
    SUM(any_emvar::INT)>0 AS true_any_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    source_table,
    tag_snp;

 * postgresql://mr_root:***@localhost:5433/scratch
Done.
430495 rows affected.


[]

Now let's summarize how many combos have any emvars.

In [13]:
%%sql
SELECT
    source_table,
    COUNT(*) AS total_tag_snps,
    SUM(true_any_emvar_count::INT) AS any_emvar_tag_snps
FROM 
    tags
GROUP BY 
    source_table;

 * postgresql://mr_root:***@localhost:5433/scratch
3 rows affected.


[('AFR', 86276, 9035), ('ASN', 102101, 24373), ('EUR', 242118, 54049)]

This table, "buddies", reports, for each combination of source_table, ld_buddy, whether that combo has any emvars.

In [14]:
%%sql
--DROP table buddies;
CREATE TABLE buddies AS
SELECT
    source_table,
    ld_buddy,
    SUM(any_emvar::INT)>0 AS true_any_emvar_count
FROM
    stacked_table_deduplicated
GROUP BY
    source_table,
    ld_buddy;

 * postgresql://mr_root:***@localhost:5433/scratch
Done.
3142265 rows affected.


[]

2025-03-18 13:57:47.704 EDT [847201] LOG:  checkpoint complete: wrote 1076 buffers (6.6%); 0 WAL file(s) added, 0 removed, 33 recycled; write=235.653 s, sync=0.622 s, total=236.459 s; sync files=41, longest=0.401 s, average=0.016 s; distance=539533 kB, estimate=539533 kB; lsn=11D/67C56B70, redo lsn=11D/551F81D8


Now let's summarize how many combos have any emvars.

In [15]:
%%sql
SELECT
    source_table,
    COUNT(*) AS total_buddies,
    SUM(true_any_emvar_count::INT) AS any_emvar_buddies
FROM 
    buddies
GROUP BY 
    source_table;

 * postgresql://mr_root:***@localhost:5433/scratch
3 rows affected.


[('AFR', 468252, 6838), ('ASN', 995502, 14666), ('EUR', 1678511, 23951)]

# Dump stacked table

In [16]:
%%sql
COPY stacked_table_deduplicated
TO '/home/mcn26/varef/data/GWAS_malin/gwas_subset_malin_preds.tsv'
WITH DELIMITER E'\t' HEADER;

 * postgresql://mr_root:***@localhost:5433/scratch


2025-03-18 13:58:51.020 EDT [847201] LOG:  checkpoint starting: time


16412995 rows affected.


[]

# Shut down server

In [17]:
!pg_ctl -D ~/palmer_scratch/db stop

waiting for server to shut down....

2025-03-18 13:59:10.480 EDT [847200] LOG:  received fast shutdown request
2025-03-18 13:59:10.483 EDT [847200] LOG:  aborting any active transactions
2025-03-18 13:59:10.483 EDT [847325] FATAL:  terminating connection due to administrator command
2025-03-18 13:59:10.490 EDT [847200] LOG:  background worker "logical replication launcher" (PID 847206) exited with exit code 1
2025-03-18 13:59:10.494 EDT [847660] FATAL:  terminating autovacuum process due to administrator command
2025-03-18 13:59:10.494 EDT [847660] CONTEXT:  while scanning block 27367242 of relation "public.unique_id_malin_gnomad"


.

2025-03-18 13:59:11.454 EDT [847201] LOG:  checkpoint complete: wrote 162 buffers (1.0%); 0 WAL file(s) added, 0 removed, 19 recycled; write=20.291 s, sync=0.007 s, total=20.434 s; sync files=35, longest=0.002 s, average=0.001 s; distance=321373 kB, estimate=517717 kB; lsn=11D/69093F88, redo lsn=11D/68BCF8B8
2025-03-18 13:59:11.454 EDT [847201] LOG:  shutting down
2025-03-18 13:59:11.455 EDT [847201] LOG:  checkpoint starting: shutdown immediate
2025-03-18 13:59:11.587 EDT [847201] LOG:  checkpoint complete: wrote 277 buffers (1.7%); 0 WAL file(s) added, 1 removed, 0 recycled; write=0.108 s, sync=0.003 s, total=0.133 s; sync files=3, longest=0.003 s, average=0.001 s; distance=4881 kB, estimate=466434 kB; lsn=11D/69094018, redo lsn=11D/69094018
2025-03-18 13:59:11.619 EDT [847200] LOG:  database system is shut down


 done
server stopped
